platform/pulpcore/app/models/storage.py

```diff
@@ -101,300 +97,86 @@ def __call__(self, model, name):
             self.name)
 
 
-class FileSystem(Storage):
+class FileSystem(FileSystemStorage):
     """
-    Provides simplified filesystem storage.
-    The django filesystem storage is overly complex and makes incompatible assumptions
-    and has incompatible behaviors with regard to how final storage paths are calculated.
+    Modifies how Django saves files
     """
 
-    @staticmethod
-    def unlink(path):
+    def get_available_name(self, name, max_length=None):
         """
-        Delete the link at the specified path.
-        No exception is raised if the link does not exist.
+        Returns a filename if a file by that name does not exist
 
         Args:
-            path (str): Path to unlink.
-
-        Raises:
-            OSError: On error.
-        """
-        try:
-            os.unlink(path)
-        except FileNotFoundError:
-            pass
-
-    @staticmethod
-    def copy(path, content):
-        """
-        Atomic copy of the content into storage.
-        The content copied into a temporary file, then renamed.
-        This prevents creating partial files.
+            name (string): Requested file name
+            max_length (int): Maximum length of the filename. Not used in this implementation.
 
-        Args:
-            path(str): The storage path.
-            content (File): Source file object.
+        Returns:
+            Name of the file.
 
         Raises:
-            FileExistsError: File already stored.
-            OSError: On error.
+            OSError if a file with requested name already exists.
         """
-        tmp_path = '.'.join((path, str(uuid4()), 'tmp'))
-        try:
-            with open(tmp_path, 'wb+') as fp:
-                while True:
-                    bfr = content.read(1024000)
-                    if bfr:
-                        fp.write(bfr)
-                    else:
-                        break
-            os.link(tmp_path, path)
-        finally:
-            FileSystem.unlink(tmp_path)
-
-    @staticmethod
-    def delete_empty_dirs(path, root):
-        """
-        Delete empty directories in the specified path.
-
-        Args:
-            path (str): An absolute path to directory to delete
-            root (str): An absolute path to directory which should not be removed even if empty
-        """
-        if root == path:
-            return
-        try:
-            os.rmdir(path)
-        except OSError as e:
-            if e.errno in [errno.ENOENT, errno.ENOTEMPTY]:
-                return
-            else:
-                raise
+        if self.exists(name):
```

Comment:
This is not thread safe.

Reply:
Yes, but this method is called again later in _save() if a file exists exception is raised.

Reply:
The `Storage` interface does not document that this method should raise `FileExistsError`.  This behavior is convenient for changing the logic in _save() but inconsistent with the `Storage` interface and dangerous.  What if the `FileSystemStorage._save()` logic changes?  This could result in unpredictable behavior.